### PR TITLE
Unit testing

### PR DESCRIPTION
**Pull request by nvnyale**

### CODE REVIEW

**Review by tenyo — COMMENTED** [review]
Looks pretty good! I left some comments. Once we have this first test done it should be pretty easy to add more.
I think you can also test what happens when you pass an empty group or task name to startTaskRun (should get `apierror.ErrBadRequest`).

File: api/orchestration_movers_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/YaleSpinup/aws-go/services/iam"
+	yresource "github.com/YaleSpinup/aws-go/services/resourcegroupstaggingapi"
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done..

File: api/orchestration_movers_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	ydataSync "github.com/YaleSpinup/datasync-api/datasync"
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done..

File: api/orchestration_movers_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/YaleSpinup/aws-go/services/iam"
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done..

File: api/orchestration_movers_test.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/datasync"
+	"github.com/aws/aws-sdk-go/service/datasync/datasynciface"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi/resourcegroupstaggingapiiface"
+	"github.com/stretchr/testify/assert"
+)
+
+//datasyncClient
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Removed comment.

File: api/orchestration_movers_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+type mockDatasync struct {
+	datasynciface.DataSyncAPI
+	t   *testing.T
+	err error
+}
+
+//rgClient
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Removed..

File: api/orchestration_movers_test.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+type mockrgClient struct {
+	resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
+	t   *testing.T
+	err error
+}
+
+// We dont want to start task so we mocked it
+func (d *mockDatasync) StartTaskExecutionWithContext(ctx context.Context, input *datasync.StartTaskExecutionInput, opts ...request.Option) (*datasync.StartTaskExecutionOutput, error) {
+	fmt.Println("mocked StartTaskExecutionWithContext")
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Removed.

File: api/orchestration_movers_test.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	if d.err != nil {
+		return nil, d.err
+	}
+	fmt.Print(ctx, &input, opts)
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done..

File: api/orchestration_movers_test.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	out := &datasync.StartTaskExecutionOutput{
+		TaskExecutionArn: aws.String("arn:aws:datasync:us-east-1:516855177326:task/task-05cd6f77d7b5d15ac/execution/exec-086d6c629a6bf3581"),
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Code is now updated to 012345678901

File: api/orchestration_movers_test.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	return out, nil
+}
+
+var is_running = false
+
+// It returns a task with RUNNING / AVAIABLE bases on is_running
+func (d *mockDatasync) DescribeTaskWithContext(ctx context.Context, input *datasync.DescribeTaskInput, opts ...request.Option) (*datasync.DescribeTaskOutput, error) {
+	fmt.Println("mocked DescribeTaskWithContext")
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Initially i have written this for troubleshooting, now removed it. Thanks.

File: api/orchestration_movers_test.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		return &datasync.DescribeTaskOutput{
+			Name:    aws.String("name1"),
+			Status:  aws.String("RUNNING"),
+			TaskArn: aws.String("arn:aws:datasync:us-east-1:516855177326:task/task-05cd6f77d7b5d15ac/execution/exec-086d6c629a6bf3581"),
+		}, nil
+	}
+	return &datasync.DescribeTaskOutput{
+		Name:    aws.String("name1"),
+		Status:  aws.String("AVAILABLE"),
+		TaskArn: aws.String("arn:aws:datasync:us-east-1:516855177326:task/task-05cd6f77d7b5d15ac/execution/exec-086d6c629a6bf3581"),
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Updated.

File: api/orchestration_movers_test.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+// It returns a task with RUNNING / AVAIABLE bases on is_running
+func (d *mockDatasync) DescribeTaskWithContext(ctx context.Context, input *datasync.DescribeTaskInput, opts ...request.Option) (*datasync.DescribeTaskOutput, error) {
+	fmt.Println("mocked DescribeTaskWithContext")
+	if is_running {
+
+		return &datasync.DescribeTaskOutput{
+			Name:    aws.String("name1"),
+			Status:  aws.String("RUNNING"),
+			TaskArn: aws.String("arn:aws:datasync:us-east-1:516855177326:task/task-05cd6f77d7b5d15ac/execution/exec-086d6c629a6bf3581"),
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done.

File: api/orchestration_movers_test.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	return &datasync.DescribeTaskOutput{
+		Name:    aws.String("name1"),
+		Status:  aws.String("AVAILABLE"),
+		TaskArn: aws.String("arn:aws:datasync:us-east-1:516855177326:task/task-05cd6f77d7b5d15ac/execution/exec-086d6c629a6bf3581"),
+	}, nil
+}
+
+//Get tags
+func (r *mockrgClient) GetResourcesWithContext(ctx context.Context, input *resourcegroupstaggingapi.GetResourcesInput, opts ...request.Option) (*resourcegroupstaggingapi.GetResourcesOutput, error) {
+	fmt.Println("mocked GetResourcesWithContext")
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Removed.

File: api/orchestration_movers_test.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func newmockrgClient(t *testing.T, err error) resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI {
+	return &mockrgClient{
+		t:   t,
+		err: err,
+	}
+}
+
+func Test_StartRun_Runs_When_Status_Available(T *testing.T) {
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Updated to small t.

File: api/orchestration_movers_test.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+		rgClient: yresource.ResourceGroupsTaggingAPI{
+			Service: newmockrgClient(T, nil),
+		},
+	}
+	resp, err := o.startTaskRun(nil, "group1", "name1")
+	assert.NoError(T, err, "no error")
+	assert.Equal(T, "exec-086d6c629a6bf3581", resp)
+}
+
+func Test_StartRun_Dose_Not_Run_If_its_running(T *testing.T) {
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done..

File: api/orchestration_movers_test.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+			DefaultKMSKeyId: "",
+		},
+		iamClient: iam.IAM{},
+		rgClient: yresource.ResourceGroupsTaggingAPI{
+			Service: newmockrgClient(T, nil),
+		},
+	}
+	resp, err := o.startTaskRun(nil, "group1", "name1")
+	assert.NoError(T, err, "no error")
+	assert.Equal(T, "exec-086d6c629a6bf3581", resp)
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: I usually use the assert syntax as it is common in every testing syntax in other languages. If you think it is good to get rid of this, please let me know, I can also write go assert struct which will look like this and we don't need to code stretcher/testify/assert every time.

File: api/orchestration_movers_test.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+			DefaultKMSKeyId: "",
+		},
+		iamClient: iam.IAM{},
+		rgClient: yresource.ResourceGroupsTaggingAPI{
+			Service: newmockrgClient(T, nil),
+		},
+	}
+	resp, err := o.startTaskRun(nil, "group1", "name1")
+	assert.Error(T, err, "Conflict: another datasync mover task is already running")
+	assert.Equal(T, "", resp)
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_movers_test.go]: Done

**Review by tenyo — APPROVED** [review]
Looks good 👍
